librad/src/git/storage.rs

```diff
@@ -380,25 +184,16 @@ impl AsRef<Storage> for Storage {
     }
 }
 
-struct ReferenceNames<'a> {
-    iter: git2::References<'a>,
+impl AsRef<read::Storage> for Storage {
+    fn as_ref(&self) -> &read::Storage {
+        &self.inner
+    }
 }
 
-impl<'a> Iterator for ReferenceNames<'a> {
-    type Item = Result<ext::RefLike, Error>;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let names = self.iter.names();
-        for name in names {
-            match name {
-                Err(e) => return Some(Err(e.into())),
-                Ok(name) => match ext::RefLike::try_from(name).ok() {
-                    Some(refl) => return Some(Ok(refl)),
-                    None => continue,
-                },
-            }
-        }
+impl std::ops::Deref for Storage {
```

Comment:
So, I know we use this pattern in a few places but it's also [considered an anti-pattern](https://github.com/rust-unofficial/patterns/blob/master/anti_patterns/deref.md). Removing it would result in a good bit of boilerplate. What's our general stance on this?

Reply:
It's not advisable indeed, except in YOLO circumstances. As those anti-patternists say, `*x.deref()` does not have the same type as `x`, which can yield surprises when rust decides to auto-dereference.

I think in this case the only way to avoid having to borrow explicitly at the call site is to turn `Storage` into a pair of traits.

Reply:
Why do you say a pair of traits? One of them being the read-only operations like `has_*` and the reference lookups, but what would the other one consist of?

Reply:
Well, say `trait Storage: ReadOnlyStorage`. Should I have written “couple” to denote it’s two traits? xoxo

Maybe `struct Storage` could be an impl of `ReadOnlyStorage`, which would save one trait and may make the transition easier?

Reply:
> xoxo

:kissing_closed_eyes: 

> Well, say trait Storage: ReadOnlyStorage. Should I have written “couple” to denote it’s two traits?

What I'm wondering is what do you see going into the `Storage` trait? :) The only operations left are constructors and `as_raw`. So maybe whatever the `as_raw` is doing in the rest of the codebase (where writes are happening) should be considered going into that trait?

> Maybe struct Storage could be an impl of ReadOnlyStorage, which would save one trait and may make the transition easier?

That makes sense. So just to check we're on the same page -- instead of things taking `S: AsRef<ReadOnly>` they would take `S: ReadOnlyStorage`, right?

Reply:
> What I'm wondering is what do you see going into the Storage trait?

Instead of `as_raw`, we should probably strive for the respective operations be captured in either their own traits, or as methods on a rw storage. This will become more pressing in a hybrid git2/gitoxide world I suppose. It’s a lot of work for internal interfaces, which is why I didn’t do it. Perhaps we could capture `as_raw` as a trait in the meantime, because that can be made crate-visible only.

Btw, I’ll also need to take another look at whether none of the git2 types returned from the ro storage can be used mutably.

> instead of things taking S: AsRef<ReadOnly> they would take S: ReadOnlyStorage, right?
 
 Uhm, `AsRef` is actually fine, if you’re fine with calling `as_ref` everywhere. What boilerplate did you see arise without `Deref`?

Reply:
> Instead of as_raw, we should probably strive for the respective operations be captured in either their own traits

Ya, I'm with you :+1: 

> Btw, I’ll also need to take another look at whether none of the git2 types returned from the ro storage can be used mutably.

Oh that's a great point :)

> Uhm, AsRef is actually fine, if you’re fine with calling as_ref everywhere. What boilerplate did you see arise without Deref?

Actually you're right. The boilerplate was due to methods like `has_urn` being used on `Storage` directly.

I suppose one other alternative to the trait is having top-level functions, for example:
```rust
// storage/read.rs
pub fn has_urn(storage: &S, urn: &Urn) -> Result<bool, Error> 
where S: AsRef<ReadOnly>,
{
    let storage = storage.as_ref();
    /* etc */
}
```

That would mean making a lot of changes to method calls in the codebase though. Thought I'd throw it out there nonetheless.

Reply:
> top-level functions

Yeah, I mean, that’s the pattern used for all the higher-level effects on storage. I thought that one day I’ll add capability traits instead of passing around the concrete `Storage`, but well...